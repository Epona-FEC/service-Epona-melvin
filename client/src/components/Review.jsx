/* eslint-disable jsx-a11y/img-redundant-alt */
import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';

const Review = ({ product, review, reviewer, renderStars }) => {
  const { name, id } = product;
  const { avatar, username } = reviewer;
  const { date, score, body } = review;

  return (
    <div className="review">
      {/* Possible Component to Refactor */}
      <div className='review-info'>
        <img className='avatar' src={avatar} alt={`${username}'s avatar`} />
        <a className='reviewer-name review-link' href={`/users/${username}`}>
          { username }
        </a>
        <em className='review-date'>{ moment(date).format('MMM DD, YYYY')}</em>
      </div>
      {/* ====== */}
      <div className='review-body'>
        <div className='review-data'>
          <div className='review-score'>
            { score }
          </div>
          <span className='review-text'>{body}</span>
        </div>
        <img className='review-photo' src={review.photoUrl} alt={`${username} added a photo of their purchase`} />
        {/* Shopowner Div HERE  TODO: REPLY */}
        <div className='review-product-data'>
          <span>Purchased item:</span>
          <div>
            <img className='product-icon' src={product.photoUrl} alt={name} />
            <a href={`/listing/${id}`} className='review-link'>
              { name }
            </a>
          </div>
        </div>
      </div>
    </div>
  );
};
Review.propTypes = {

  product: PropTypes.shape({
    name: PropTypes.string,
    id: PropTypes.number,
    photoUrl: PropTypes.string,
  }),
  review: PropTypes.shape({
    date: PropTypes.instanceOf(Date),
    score: PropTypes.number,
    body: PropTypes.string,
    photoUrl: PropTypes.string,
  }),
  reviewer: PropTypes.shape({
    avatar: PropTypes.string,
    username: PropTypes.string,
  }),
  renderStars: PropTypes.function
};

Review.defaultProps = {
  product: {
    name: 'A Cool Product',
    id: 1,
    photoUrl: 'https://picsum.photos/id/237/200/300',
  },
  review: {
    date: new Date('Tue Mar 17 2020 16:08:19 GMT-0700 (Pacific Daylight Time)'),
    score: 5,
    body: 'This a test',
    photoUrl: 'https://picsum.photos/id/237/200/300',
  },
  reviewer: {
    avatar: 'http://www.gravatar.com/avatar/?d=identicon',
    username: 'Melvin',
  },

};

export default Review;
