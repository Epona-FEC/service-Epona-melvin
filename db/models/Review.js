module.exports = (sequelize, Sequelize) => {
  const Review = sequelize.define('Review', {
    id: {
      type: Sequelize.INTEGER,
      primaryKey: true,
      autoIncrement: true,
    },
    body: {
      type: Sequelize.TEXT,
    },
  });
  return Review;
};
