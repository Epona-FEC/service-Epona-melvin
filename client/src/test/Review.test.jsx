/* eslint-disable no-undef */
import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import Review from '../components/Review.jsx';

Enzyme.configure({ adapter: new Adapter() });

describe('Review', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Review />);
  });

  it('Should shallow', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renderStars should return score', () => {
    const getScore = (score) => score;
    wrapper.setProps({ renderStars: getScore });
    const scoreProp = wrapper.props('reviewScore').score;
    const defaultScore = getScore(scoreProp);
    expect(defaultScore).toBe(scoreProp);
  });

  it('renderStars function should return array of 5 children review-score', () => {
    const renderFiveSpans = () => {
      const arr = [];
      for (let i = 0; i < 5; i += 1) {
        arr.push(<span>i</span>);
      }
      return arr;
    };
    wrapper.setProps({ renderStars: renderFiveSpans });
    expect(wrapper.find('.review-score').props().children).toHaveLength(5);
  });
});
