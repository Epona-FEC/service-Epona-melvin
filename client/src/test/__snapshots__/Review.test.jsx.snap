// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Review Should shallow 1`] = `
<div
  className="review"
>
  <div
    className="review-info"
  >
    <img
      alt="Melvin's avatar"
      className="avatar"
      src="http://www.gravatar.com/avatar/?d=identicon"
    />
    <a
      className="reviewer-name review-link"
      href="/users/Melvin"
    >
      Melvin
    </a>
    <em
      className="review-date"
    >
      Mar 17, 2020
    </em>
  </div>
  <div
    className="review-body"
  >
    <div
      className="review-data"
    >
      <div
        className="review-score"
      />
      <span
        className="review-text"
      >
        This a test
      </span>
    </div>
    <img
      alt="Melvin added a photo of their purchase"
      className="review-photo"
      src="https://picsum.photos/id/237/200/300"
    />
    <div
      className="review-product-data"
    >
      <span>
        Purchased item:
      </span>
      <div>
        <img
          alt="A Cool Product"
          className="product-icon"
          src="https://picsum.photos/id/237/200/300"
        />
        <a
          className="review-link"
          href="/listing/1"
        >
          A Cool Product
        </a>
      </div>
    </div>
  </div>
</div>
`;
